---
title: "rna_seq_tutorial"
author: "Shaleigh Smith"
date: "4/29/2019"
output: html_document
---



---

RNA-Seq Tutorial

---

This RNA-Seq tutorial is designed to give you an insight into how to download data from the Sequence Read Archive (SRA), run quality control, trimming, and alignments on the command line, and analyze the results using DESeq2. 

This tutorial was made to be run on a high-throughput computing cluster with SLURM job manager. This can be done on your own system if the necessary software is installed. Please note that some of the steps are computationally heavy and will take a longer time on your local system. 

---

We will be working with lung adenocarcinoma RNA-Seq data obtained from:

Seo JS, Ju YS, Lee WC, Shin JY et al. The transcriptional landscape and mutational profile of lung adenocarcinoma. Genome Res 2012 Nov;22(11):2109-19. PMID: 22975805
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40419

This includes paired normal and tumor samples. We will only be using a small subset of the samples, though the entire dataset is available for download

---

First we will be creating a custom script to:

1. Download the RNA-Seq Datasets & Human Genome
2. Check the quality of the read
3. Trim the reads for low quality and adapters (if needed)
4. Align the reads to the reference genome 
5. Convert and sort file formats
6. Create a feature counts matrix 

---


```{bash}

# Login to the cluster and create a new directory for this tutorial
# This is where all of the files will be located 
mkdir rna_seq_tutorial

# Open up a text editor in the new file (I use nano) and we will begin to write the first script
cd rna_seq_tutorial
nano

```

---

Create a submitter script. This will submit you script for you and run the following tasks on each dataset in parallel.

See submitter.txt in file (or below)

---

```{bash}

#!/bin/bash
#SBATCH --job-name=job_submitter      ### name of job
#SBATCH --nodes=1                     ### number of nodes to run on
#SBATCH --mem=200MB                   ### memory
#SBATCH --time=1:00:00                ### time 
#SBATCH --error=job_sub_error.txt     ### output error file
#SBATCH --output=job_sub_stdout.txt   ### output standard file

sbatch --array=1 rna_seq_script.sh LC_C1 ERX140427
sbatch --array=1 rna_seq_script.sh LC_C2 ERX140428
sbatch --array=1 rna_seq_script.sh LC_C3 ERX140429
sbatch --array=1 rna_seq_script.sh LC_C5 ERX140431
sbatch --array=1 rna_seq_script.sh LC_C1_nor ERX140350
sbatch --array=1 rna_seq_script.sh LC_C2_nor ERX140351
sbatch --array=1 rna_seq_script.sh LC_C3_nor ERX140352
sbatch --array=1 rna_seq_script.sh LC_C5_nor ERX140353

# Save this as submitter.sh

```


```{bash}

# Create a header for your rna script

#!/bin/bash
#SBATCH --job-name=rna_seq                   ### Job name
#SBATCH --mail-type=END,FAIL                 ### Mail events (NONE, BEGIN, END, FAIL, ALL)
#SBATCH --mail-user=name.last@school.org     ### Where to send mail
#SBATCH --ntasks=4                           ### Run on a 4 CPU
#SBATCH --mem=64gb                           ### Job memory request
#SBATCH --time=10:00:00                      ### Time limit hrs:min:sec
#SBATCH --output=./rna_seq_%j.log            ### Standard output and error log
#SBATCH -p cpu_short                         ### Node 

```


```{bash}

# Download the Human Genome (hg38) 
wget ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg38/Homo_sapiens_UCSC_hg38.tar.gz
tar -zxvf Homo_sapiens_UCSC_hg38.tar.gz

```


```{bash}

# Import modules
module load sratoolkit/2.9.1
module load fastqc/0.11.7
module load trimgalore/0.5.0
module load python/cpu/2.7.15-ES ### CutAdapt is hidden in here
module load bbmap/38.25
module load samtools/1.9
module load subread/1.6.3

```

```{bash}

# Download Data from the SRA
fastq-dump ${1} --split-files --gzip -O ./

# Remove fastq-dump directory
rm -r ~/ncbi

# Rename files
mv ./${1}*fastq.gz ./${2}.fastq.gz

```

```{bash}

# Trim and run FASTQC
trim_galore --q 30 \
--phred33 \
-o ./ \
--paired \
--fastqc \
./${2}_1.fastq.gz \
./${2}_2.fastq.gz

```

```{bash}

# Align reads to reference genome
bbmap.sh \
-Xmx26G \
ref=./hg38/Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.fa \
in=./${2}_val_1.fq.gz \
in2=./${2}_val_2.fq.gz \
outm=./${2}.sam \
minid=0.90 \
ambiguous=random \
nodisk \

```

```{bash}

# Convert SAM file to BAM file and sort
samtools view -S -b ${2}.sam > ${2}..bam
samtools sort ${2}.bam -o ${2}_sorted.bam

```


There is a large variety of flags that can be used for feature counts. In this case we will be more stringent with the following parameters.

The input is specified as reversely stranded (-s 2) and paired end (-p). The fragment length is checked (-P) and only the fragments that have both ends successfully aligned are counted (-B). Chimeric fragments are ignored (-C) as are duplicates (--ignoreDup). Finally, only primary alignments are counted (--primary) and the counts are based on gene_id (-g).


```{bash}

# Create a feature counts matrix
featureCounts -s 2 -p -B -C -P --ignoreDup --primary -a ./hg38/Homo_sapiens/UCSC/hg38/Annotation/Genes.gencode/genes.gtf -g gene_id -o ${2}_feature_counts ./${2}_sorted.bam

```

```{bash}

# The final script should look like this:


#!/bin/bash
#SBATCH --job-name=rna_seq                   ### Job name
#SBATCH --mail-type=END,FAIL                 ### Mail events (NONE, BEGIN, END, FAIL, ALL)
#SBATCH --mail-user=name.last@school.org     ### Where to send mail
#SBATCH --ntasks=4                           ### Run on a 4 CPU
#SBATCH --mem=64gb                           ### Job memory request
#SBATCH --time=10:00:00                      ### Time limit hrs:min:sec
#SBATCH --output=./rna_seq_%j.log            ### Standard output and error log
#SBATCH -p cpu_short                         ### Node 

# Download the Human Genome (hg38) 
wget ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg38/Homo_sapiens_UCSC_hg38.tar.gz
tar -zxvf Homo_sapiens_UCSC_hg38.tar.gz

# Import modules
module load sratoolkit/2.9.1
module load fastqc/0.11.7
module load trimgalore/0.5.0
module load python/cpu/2.7.15-ES ### CutAdapt is hidden in here
module load bbmap/38.25
module load samtools/1.9
module load subread/1.6.3

# Download Data from the SRA
fastq-dump ${1} --split-files --gzip -O ./

# Remove fastq-dump directory
rm -r ~/ncbi

# Rename files
mv ./${1}*fastq.gz ./${2}.fastq.gz

# Trim and run FASTQC
trim_galore --q 30 \
--phred33 \
-o ./ \
--paired \
--fastqc \
./${2}_1.fastq.gz \
./${2}_2.fastq.gz

# Align reads to reference genome
bbmap.sh \
-Xmx26G \
ref=./hg38/Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.fa \
in=./${2}_val_1.fq.gz \
in2=./${2}_val_2.fq.gz \
outm=./${2}.sam \
minid=0.90 \
ambiguous=random \
nodisk \

# Convert SAM file to BAM file and sort
samtools view -S -b ${2}.sam > ${2}..bam
samtools sort ${2}.bam -o ${2}_sorted.bam

# Create a feature counts matrix
featureCounts -s 2 -p -B -C -P --ignoreDup --primary -a ./hg38/Homo_sapiens/UCSC/hg38/Annotation/Genes.gencode/genes.gtf -g gene_id -o ${2}_feature_counts ./${2}_sorted.bam

```


---

# Save the script as rna_seq_script.sh

---

```{bash}

# To run your submitter (this will submit all 8 scripts, one for each dataset)
./submitter.sh

```


---

Check your output. It should include:




---

The downstream analysis will be done in R using DESeq2. 

---


```{r}

# Import feature count matrices

```


---


---


---


---


---

---