---
title: "rna_seq_tutorial"
author: "Shaleigh Smith"
date: "4/29/2019"
output:
  html_document: default
  pdf_document: default
---



---

RNA-Seq Tutorial

---

This RNA-Seq tutorial is designed to give you an insight into how to download data from the Sequence Read Archive (SRA), run quality control, trimming, and alignments on the command line, and analyze the results using DESeq2. 

This tutorial was made to be run on a high-throughput computing cluster with SLURM job manager. This can be done on your own system if the necessary software is installed. Please note that some of the steps are computationally heavy and will take a longer time on your local system. 

---

We will be working with lung adenocarcinoma RNA-Seq data obtained from:

Seo JS, Ju YS, Lee WC, Shin JY et al. The transcriptional landscape and mutational profile of lung adenocarcinoma. Genome Res 2012 Nov;22(11):2109-19. PMID: 22975805
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40419

This includes paired normal and tumor samples. We will only be using a small subset of the samples, though the entire dataset is available for download. The first part of this tutorial will explain how to write the script for all data pre-processing. The second part is a simple workflow through DESeq2 to identify up and down regulated genes between normal and tumor samples.

---

First we will be creating a custom script to:

1. Download the RNA-Seq Datasets & Human Genome
2. Check the quality of the read
3. Trim the reads for low quality and adapters (if needed)
4. Align the reads to the reference genome 
5. Convert and sort file formats
6. Create a feature counts matrix 

---

Login to the cluster and create a new directory for this tutorial.
This is where all of the files will be located: 

#### mkdir rna_seq_tutorial


---

Open up a text editor in the new file (I use nano) and we will begin to write the first script

#### cd rna_seq_tutorial
#### nano


---

Create a submitter script. This will submit you script for you and run the following tasks on each dataset in parallel.

See submitter.txt in file (or below)

---

#### #!/bin/bash  
#### #SBATCH --job-name=job_submitter      ### name of job  
#### #SBATCH --nodes=1                     ### number of nodes to run on  
#### #SBATCH --mem=200MB                   ### memory  
#### #SBATCH --time=1:00:00                ### time   
#### #SBATCH --error=job_sub_error.txt     ### output error file  
#### #SBATCH --output=job_sub_stdout.txt   ### output standard file  
 
 

####sbatch --array=1 rna_seq_script.sh ERX140427  LC_C1

####sbatch --array=1 rna_seq_script.sh ERX140428  LC_C2

####sbatch --array=1 rna_seq_script.sh ERX140429  LC_C3

####sbatch --array=1 rna_seq_script.sh ERX140431  LC_C5

####sbatch --array=1 rna_seq_script.sh ERX140350  LC_C1_nor

####sbatch --array=1 rna_seq_script.sh ERX140351  LC_C2_nor

####sbatch --array=1 rna_seq_script.sh ERX140352  LC_C3_nor

####sbatch --array=1 rna_seq_script.sh ERX140353  LC_C5_nor

---

Save this as submitter.sh

---

Open a new empty text file and treate a header for your rna script:

#### #!/bin/bash
 
#### #SBATCH --job-name=rna_seq                   ### Job name
 
#### #SBATCH --mail-type=END,FAIL                 ### Mail events (NONE, BEGIN, END, FAIL, ALL)
 
#### #SBATCH --mail-user=name.last@school.org     ### Where to send mail
 
#### #SBATCH --ntasks=4                           ### Run on a 4 CPU
 
#### #SBATCH --mem=64gb                           ### Job memory request
 
#### #SBATCH --time=12:00:00                      ### Time limit hrs:min:sec
 
#### #SBATCH --output=./rna_seq_%j.log            ### Standard output and error log
 
#### #SBATCH -p cpu_short                         ### Node 

---


---

Download the Human Genome (hg38) 

#### wget ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg38/Homo_sapiens_UCSC_hg38.tar.gz
#### tar -zxvf Homo_sapiens_UCSC_hg38.tar.gz

---


---

Import modules

#### module load sratoolkit/2.9.1
#### module load fastqc/0.11.7
#### module load trimgalore/0.5.0
#### module load python/cpu/2.7.15-ES ## CutAdapt is hidden in here
#### module load bbmap/38.25
#### module load samtools/1.9
#### module load subread/1.6.3

---


---

Download Data from the SRA

#### fastq-dump ${1} --split-files --gzip -O ./

---

Remove fastq-dump directory

#### rm -r ~/ncbi

---

Rename files

#### mv ./ ${1}_1*fastq.gz     ./ ${2}_1.fastq.gz


#### mv ./ ${1}_2*fastq.gz     ./ ${2}_2.fastq.gz

---


---

Trim and run FASTQC

#### trim_galore --q 30  --phred33  -o ./  --paired  --fastqc  ./ ${2}_1.fastq.gz  ./ ${2}_2.fastq.gz

---


---

Align reads to reference genome

#### bbmap.sh -Xmx26G ref=./Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.fa 
#### in= ./ ${2}_val_1.fq.gz  in2= ./ ${2}_val_2.fq.gz  outm= ./ ${2}.sam  minid=0.90  ambiguous=random  nodisk 

---


---

Convert SAM file to BAM file and sort

#### samtools view -S -b ${2}.sam > ${2}.bam
#### samtools sort ${2}.bam -o ${2}_sorted.bam

---


---

There is a large variety of flags that can be used for feature counts. In this case we will be more stringent with the following parameters.

The input is specified as reversely stranded (-s 2) and paired end (-p). The fragment length is checked (-P) and only the fragments that have both ends successfully aligned are counted (-B). Chimeric fragments are ignored (-C) as are duplicates (--ignoreDup). Finally, only primary alignments are counted (--primary) and the counts are based on gene_id (-g).

---


---

Create a feature counts matrix

#### featureCounts -s 2 -p -B -C -P  --ignoreDup  --primary  -a  ./Homo_sapiens/UCSC/hg38/Annotation/Genes.gencode/genes.gtf  -g gene_id  -o $ {2} -feature-counts   ./${2}_sorted.bam

---

Save the script as rna_seq_script.sh


---


---

The final script should look include all of these commands. See 'rna_seq_script.sh' in the tutorial file for the final example. 

---

---


To run your submitter (this will submit all 8 scripts, one for each dataset)

#### ./submitter.sh

---

Check the output. It should include these files for each dataset:

1. LC_C1_1.fastq.gz
2. LC_C1_1.fastq.gz_trimming_report.txt
3. LC_C1_1_val_1_fastqc.html
4. LC_C1_1_val_1_fastqc.zip
5. LC_C1_1_val_1.fq.gz
6. LC_C1_2.fastq.gz
7. LC_C1_2.fastq.gz_trimming_report.txt
8. LC_C1_2_val_2_fastqc.html
9. LC_C1_2_val_2_fastqc.zip
10. LC_C1_2_val_2.fq.gz
11. LC_C1.bam
12. LC_C1_feature_counts
13. LC_C1_feature_counts.summary

---

Copy the feature counts files to your local directory. You can do this using a command or an application like FileZilla. 

---

---

---

The downstream analysis will be done in R using DESeq2. Open up a new script in R - I prefer using an R Markdown file so that you can visualize the results for each step. 

The general workflow is as follows:

1. Import count matrices
2. Normalization
3. Transformation
4. LFC shrinkage
5. Data exploration (sample relatedness)
6. Difference gene expression analysis

---

```{r message=FALSE}

# Load Libraries
library(tidyverse)
library(plyr)
library(dplyr)
library(DESeq2)
#library(purrr)
library(ComplexHeatmap)
library(circlize)
library(ggplot2)
library(ggfortify)
#library(vsn)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(genefilter)
library(biomaRt)
library(data.table)
#library(IHW)
#library(VennDiagram)
#library(gridExtra)

# Not all of these are required - this depends on how you choose to visualize your results 

```


```{r}

### Import & clean gene count files

count_1 <- read.table("./data/LC_C1_nor_feature_counts", header=TRUE) 
colnames(count_1)[c(1, 7)] <- c("gene_id", "LC_C1_nor")
count_1 <- dplyr::select(count_1, gene_id, LC_C1_nor)
count_1$gene_id <- substring(count_1$gene_id, 1, 15)

count_2 <- read.table("./data/LC_C1_feature_counts", header=TRUE) 
colnames(count_2)[c(1, 7)] <- c("gene_id", "LC_C1")
count_2 <- dplyr::select(count_2, gene_id, LC_C1)
count_2$gene_id <- substring(count_2$gene_id, 1, 15)

count_3 <- read.table("./data/LC_C2_nor_feature_counts", header=TRUE) 
colnames(count_3)[c(1, 7)] <- c("gene_id", "LC_C2_nor")
count_3 <- dplyr::select(count_3, gene_id, LC_C2_nor)
count_3$gene_id <- substring(count_3$gene_id, 1, 15)

count_4 <- read.table("./data/LC_C2_feature_counts", header=TRUE) 
colnames(count_4)[c(1, 7)] <- c("gene_id", "LC_C2")
count_4 <- dplyr::select(count_4, gene_id, LC_C2)
count_4$gene_id <- substring(count_4$gene_id, 1, 15)

count_5 <- read.table("./data/LC_C3_nor_feature_counts", header=TRUE) 
colnames(count_5)[c(1, 7)] <- c("gene_id", "LC_C3_nor")
count_5 <- dplyr::select(count_5, gene_id, LC_C3_nor)
count_5$gene_id <- substring(count_5$gene_id, 1, 15)

count_6 <- read.table("./data/LC_C3_feature_counts", header=TRUE) 
colnames(count_6)[c(1, 7)] <- c("gene_id", "LC_C3")
count_6 <- dplyr::select(count_6, gene_id, LC_C3)
count_6$gene_id <- substring(count_6$gene_id, 1, 15)

count_7 <- read.table("./data/LC_C5_nor_feature_counts", header=TRUE) 
colnames(count_7)[c(1, 7)] <- c("gene_id", "LC_C5_nor")
count_7 <- dplyr::select(count_7, gene_id, LC_C5_nor)
count_7$gene_id <- substring(count_7$gene_id, 1, 15)

count_8 <- read.table("./data/LC_C5_feature_counts", header=TRUE) 
colnames(count_8)[c(1, 7)] <- c("gene_id", "LC_C5")
count_8 <- dplyr::select(count_8, gene_id, LC_C5)
count_8$gene_id <- substring(count_8$gene_id, 1, 15)

# You can also create a function to do this if you have a large amount of files 

```

```{r}

### Join all gene count files into one data frame
feature_counts <- join_all(list(count_1, count_3, count_5, count_7, count_2, count_4, 
                       count_6, count_8), by = "gene_id", type = "full")
# Write out table for later analysis
write.table(feature_counts, file = "feature_counts.txt", quote=FALSE, sep='\t', row.names = FALSE)

```


---

```{r}

# Read in feature counts 
deseq_counts <- read.table("feature_counts.txt", header=TRUE, row.names=1) 
as_tibble(deseq_counts)

# Create DeSeq dataset 
samples <- data.frame(row.names = c("LC_C1_nor", "LC_C2_nor", "LC_C3_nor", "LC_C5_nor",
                                    "LC_C1", "LC_C2", "LC_C3", "LC_C5"),
                      sample = as.factor(c(rep("normal", 4), 
                                           rep("tumor", 4))))
deseq_df <- DESeqDataSetFromMatrix(countData = deseq_counts, colData = samples, 
                                   design = ~ sample)

# Remove genes that do not have counts greater than 2 in at least 2 of the datasets (columns)
deseq_df <- deseq_df[rowSums(counts(deseq_df) >= 2) >= 2] 

# Confirm that all samples are labelled correctly 
as.data.frame(colData(deseq_df))

```

```{r}

### Exploratory data analysis of DESeq matrix with transformation 

# Estimate size factors
# The size factor is the median ratio of the sample over a pseudosample: for each gene, the geometric mean of all samples
deseq_eda <- estimateSizeFactors(deseq_df)

# Apply regularized-logarithm transformation
rld <- rlog(deseq_eda, blind = FALSE)

# Apply variance stabilizing transformation
vsd <- vst(deseq_eda, blind = FALSE)

# Create new data frame three normalization methods for all samples
deseq_eda <- bind_rows(
  as_data_frame(log2(counts(deseq_eda, normalized=TRUE)[, (1:8)])) %>%
         mutate(transform = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, (1:8)]) %>% mutate(transform = "vst"),
  as_data_frame(assay(rld)[, (1:8)]) %>% mutate(transform = "rlog"))
  
# Compare transformation visually 
ggplot(deseq_eda, aes(x = LC_C1, y = LC_C1_nor)) + 
  geom_bin2d(bins = 40) +
  coord_fixed() + 
  facet_grid( . ~ transform) +
  scale_fill_continuous(limits = c(0, 2000), breaks = seq(0, 2000, by = 500)) +
  ggtitle("LC_C1 vs. LC_C1_nor") + theme(plot.title = element_text(hjust = 0.5)) 

ggplot(deseq_eda, aes(x = LC_C1, y = LC_C2)) + 
  geom_bin2d(bins = 40) +
  coord_fixed() + 
  facet_grid( . ~ transform) +
  scale_fill_continuous(limits = c(0, 2000), breaks = seq(0, 2000, by = 500)) +
  ggtitle("LC_C1 vs. LC_C2") + theme(plot.title = element_text(hjust = 0.5))

ggplot(deseq_eda, aes(x = LC_C1_nor, y = LC_C2_nor)) + 
  geom_bin2d(bins = 40) +
  coord_fixed() + 
  facet_grid( . ~ transform) +
  scale_fill_continuous(limits = c(0, 2000), breaks = seq(0, 2000, by = 500)) +
  ggtitle("LC_C1_nor vs. LC_C2_nor") + theme(plot.title = element_text(hjust = 0.5))


```

```{r}

### Exploratory data analysis of DESeq matrix with sample comparison 
# This will asses overall similarity between samples
# As shown in the above plots this similarity might not be as expected

# Calculate the euclidean distance between samples
deseq_distance <- dist(t(assay(rld)))
deseq_distance

# Create annotation for the heatmap
deseq_annotation <- as.data.frame(as.factor(c(rep("normal", 4), 
                                              rep("tumor", 4))))
colnames(deseq_annotation)[1] <- c("sample")
deseq_ha <- HeatmapAnnotation(df = deseq_annotation, 
                              col = list(sample = c("normal" = "darkred", 
                                                    "tumor" = "darkblue")))

# Make dataframe a matrix for heatmpap
deseq_heatmap <- as.matrix(deseq_distance)

# Visualize with heatmap 
deseq_heat <- Heatmap(deseq_heatmap,
                   top_annotation = deseq_ha,
                   show_column_names = F, 
                   show_row_names = T,
                   cluster_rows = T,
                   cluster_columns = T,
                   clustering_method_columns = 'complete',
                   clustering_method_rows = "complete",
                   row_names_gp = gpar(fontsize = 8),
                   top_annotation_height = unit(1, "cm"),
                   heatmap_legend_param = list(title = "Distance"),
                   col = colorRamp2(c(0, 60), c("darkorchid4", "white")))

deseq_heat

```

```{r}

### Compare with PCA 
# visualize screeplot and autoplotted pca for reference 
deseq_pca <-prcomp(t(assay(rld)))
screeplot(deseq_pca, type='lines')
autoplot(deseq_pca)

# Create PCA object to be plotted with ggplot
plot_pca <- plotPCA(rld, intgroup = c("sample"), returnData = TRUE)

# Calculate and round variance for plotting
percentVar <- round(100 * attr(plot_pca, "percentVar"))

# Plot PCA           
deseq_pca_1 <- ggplot(plot_pca, aes(x = PC1, y = PC2, color = sample)) +
  geom_point(size = 2.5) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  scale_color_manual(values = c("darkred", "darkblue")) + 
  ggtitle("DESeq2 PCA") +
  theme(plot.title = element_text(hjust = 0.5))

deseq_pca_1

```


```{r}

# Make sure that normal is the first level in the sample factor
# This is to ensure the log2 fold change is calculated over the control (when results are called at random)
deseq_df$sample <- relevel(deseq_df$sample, "normal")

# Check to insure all samples are correct 
as.data.frame(colData(deseq_df))

# Run the DESeq analysis against raw counts 
deseq_analysis <- DESeq(deseq_df)
results(deseq_analysis)

# Call results
deseq_results <- results(deseq_analysis, contrast = c("sample", "normal", "tumor"))
deseq_results

# Review Comparisons 
mcols(deseq_results, use.names = TRUE)

### Review Summary
summary(deseq_results, alpha = 0.01)

```

```{r}

### Plot p-values for distribution check 

# Calculate row means across samples 
mean_norm <-  rowMeans(counts(deseq_analysis, normalized=TRUE)[, deseq_analysis$sample == "normal"])
mea_tumor <-  rowMeans(counts(deseq_analysis, normalized=TRUE)[, deseq_analysis$sample == "tumor"])

### Normal vs. Tumor
# Log Fold change
lfc_comp <- lfcShrink(deseq_analysis, contrast = c("sample", "normal", "tumor"))

# add row names to log fold change data frame 
lfc_comp <-  cbind(as.data.frame(lfc_comp), mean_norm, mea_tumor)

# Map gene names and ids
lfc_comp$symbol <- mapIds(org.Hs.eg.db, keys=row.names(lfc_comp), 
                             column="SYMBOL", keytype="ENSEMBL", multiVals="first")
lfc_comp$entrez <- mapIds(org.Hs.eg.db, keys=row.names(lfc_comp), 
                             column="ENTREZID", keytype="ENSEMBL", multiVals="first")

# Plot histogram to review distribution
ggplot(data = lfc_comp, mapping = aes(x = pvalue)) + geom_histogram(binwidth = 0.01)
ggplot(data = lfc_comp, mapping = aes(x = padj)) + geom_histogram(binwidth = 0.01)

```

```{r}

#### Volcano Plots

# Create up, none, down labels for color then plot

# Control 1 vs. Expression
lfc_comp <- lfc_comp %>%
  mutate(threshold = ifelse((log2FoldChange >= 0 & padj < 0.01), "up_sig",
                            ifelse((log2FoldChange <= 0 & padj < 0.01) , 
                                   "down_sig", "not_sig")))
lfc_comp$threshold[is.na(lfc_comp$threshold)] <- "not_sig"

des_vol <- ggplot(lfc_comp, aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(aes(colour = threshold), size=1) +
  ggtitle("Normal vs. Tumor with DESeq2") + 
  theme(plot.title = element_text(hjust = 0.5)) +
  ylim(0,50) + xlim(-6,6) +
  scale_colour_manual(values=c("blue", "grey", "red3"), 
                       name="Threshold 0.01",
                       breaks=c("down_sig", "not_sig", "up_sig"),
                       labels=c("Down", "None", "Up"))

des_vol

```

```{r}

#### Heatmaps with top variable genes

# Normal vs. Tumor 
# Transform with regularized log
rld <- rlog(deseq_analysis)
#assay(rld)

# Order genes then filter for the top 25
top_genes <- head(order(rowMeans(assay(rld)), decreasing = TRUE), 30)
as_tibble(top_genes)
deseq_genes <- assay(rld)[top_genes, ]

# Calculate distances from the mean for clearer heatmap
deseq_genes <- (deseq_genes - rowMeans(deseq_genes))

# Connect to ensemble database and label gene ids with gene symbol
mart <- useMart("ENSEMBL_MART_ENSEMBL","hsapiens_gene_ensembl") 
gene_symbol <- getBM(c("hgnc_symbol","ensembl_gene_id"), "ensembl_gene_id", row.names(deseq_genes), mart)

row.names(deseq_genes)[match(gene_symbol[,2], row.names(deseq_genes))] <- gene_symbol[,1]

as_tibble(deseq_genes)

# Create Annotation
sample_annotation <- as.data.frame(colData(rld)[, c("sample")])
colnames(sample_annotation)[1] <- "Sample"
sample_ha <- HeatmapAnnotation(df = sample_annotation,
                               col = list(Sample = c("normal" = "darkred", 
                                                     "tumor" = "darkblue")))

# Create Heatmap 
deseq_heat <- Heatmap(deseq_genes,
                   top_annotation = sample_ha,
                   show_column_names = F, 
                   show_row_names = T,
                   cluster_rows = T,
                   cluster_columns = T,
                   clustering_method_columns = 'complete',
                   clustering_method_rows = "complete",
                   row_names_gp = gpar(fontsize = 8),
                   top_annotation_height = unit(1, "cm"),
                   heatmap_legend_param = list(title = "Distance"),
                   col = colorRamp2(c(-1, 0, 1), c("red3", "white", "blue")),
                   column_title = "DESeq2 Normal vs. Tumor")

deseq_heat

```

```{r}

# Add column for upregulated vs. downregulated 
lfc_comp$expression <- NA
lfc_comp$expression[lfc_comp$log2FoldChange > 0 ] <- "up"
lfc_comp$expression[lfc_comp$log2FoldChange < 0 ] <- "down"
setDT(lfc_comp, keep.rownames = TRUE)[]
colnames(lfc_comp)[1] <- "gene_id"
head(lfc_comp)

# Filter out columns for significantly upregulated and downregulated genes
lfc_comp_final <- dplyr::select(lfc_comp, gene_id, symbol, log2FoldChange, padj, expression)
lfc_comp_final <- filter(lfc_comp_final, padj < 0.01)
lfc_comp_final <- lfc_comp_final[order(lfc_comp_final$padj),]
head(lfc_comp_final)

```

```{r}

# Convert gene ids to gene symbol 
mart <- useMart("ENSEMBL_MART_ENSEMBL","hsapiens_gene_ensembl") 
gene_symbol <- getBM(c("hgnc_symbol","ensembl_gene_id"), "ensembl_gene_id", row.names(deseq_analysis), mart)

row.names(deseq_analysis)[match(gene_symbol[,2], row.names(deseq_analysis))] <- gene_symbol[,1]

### Filter and order dataframes 
# State whether they are upregulated or downregulated 
# Filter out columns for significantly upregulated and downregulated genes

# Top 30 expressed genes (by p-value adjusted)
top_30 <- as.data.frame(results(deseq_analysis))
top_30 <- top_30[order(top_30$padj),]
top_30 <- head(top_30, 30)
setDT(top_30, keep.rownames = TRUE)[]
colnames(top_30)[1] <- "genes"

as_tibble(top_30)

top_30$expression <- NA
top_30$expression[top_30$log2FoldChange > 0 ] <- "up"
top_30$expression[top_30$log2FoldChange < 0 ] <- "down"
as_tibble(top_30)

```


---

```{r}

sessionInfo()

```


---


---


---

---